### PR TITLE
[libc] Add proxy header for fenv.h macro constants. #87863

### PR DESCRIPTION
**Pull request by ghost**
Hello, this addresses #87863.

### CODE REVIEW

File: libc/AOR_v20.02/math/test/mathtest.c [hunk 1]
```diff
@@ -15,7 +15,7 @@
 #include <math.h>
 #include <errno.h>
 #include <limits.h>
-#include <fenv.h>
```
Comment (lntue) [inline libc/AOR_v20.02/math/test/mathtest.c:18]: You can ignore all the files in `libc/AOR_v20.02/` folder for now.

File: libc/config/baremetal/arm/headers.txt [hunk 2]
```diff
@@ -1,7 +1,7 @@
 set(TARGET_PUBLIC_HEADERS
     libc.include.assert
     libc.include.ctype
-    libc.include.fenv
+    libc.hdr.fenv_macros
```
Comment (lntue) [inline libc/config/baremetal/arm/headers.txt:4]: Don't change `libc.include.fenv` in the `libc/config` folder, they are for registering the headers for full build modes.

File: libc/src/__support/FPUtil/FEnvImpl.h [hunk 3]
```diff
@@ -13,7 +13,7 @@
 #include "src/__support/macros/attributes.h" // LIBC_INLINE
 #include "src/__support/macros/properties/architectures.h"
 #include "src/errno/libc_errno.h"
-#include <fenv.h>
+#include "hdr/fenv_macros.h"
```
Comment (lntue) [inline libc/src/__support/FPUtil/FEnvImpl.h]: Sort the header includes.

File: libc/src/__support/FPUtil/aarch64/FEnvImpl.h [hunk 4]
```diff
@@ -17,7 +17,7 @@
 #endif
 
 #include <arm_acle.h>
-#include <fenv.h>
+#include "hdr/fenv_macros.h"
```
Comment (lntue) [inline libc/src/__support/FPUtil/aarch64/FEnvImpl.h:19]: Move this down with the `#include "src/..."`.

File: libc/src/__support/FPUtil/aarch64/fenv_darwin_impl.h [hunk 5]
```diff
@@ -17,7 +17,7 @@
 #endif
 
 #include <arm_acle.h>
-#include <fenv.h>
+#include "hdr/fenv_macros.h"
```
Comment (lntue) [inline libc/src/__support/FPUtil/aarch64/fenv_darwin_impl.h:19]: ditto

File: libc/src/__support/FPUtil/arm/FEnvImpl.h [hunk 6]
```diff
@@ -12,7 +12,7 @@
 #include "src/__support/FPUtil/FPBits.h"
 #include "src/__support/macros/attributes.h" // For LIBC_INLINE
 
-#include <fenv.h>
+#include "hdr/fenv_macros.h"
```
Comment (lntue) [inline libc/src/__support/FPUtil/arm/FEnvImpl.h:15]: Move this up to the other include group.

File: libc/src/__support/FPUtil/generic/CMakeLists.txt [hunk 7]
```diff
@@ -4,7 +4,7 @@ add_header_library(
     sqrt.h
     sqrt_80_bit_long_double.h
   DEPENDS
-    libc.include.fenv
+    libc.hdr.fenv_macros
```
Comment (lntue) [inline libc/src/__support/FPUtil/generic/CMakeLists.txt:7]: Check if the above 2 header files do not include `fenv_macros.h` then this dependency is not needed.

File: libc/src/__support/FPUtil/rounding_mode.h [hunk 8]
```diff
@@ -11,7 +11,7 @@
 
 #include "src/__support/macros/attributes.h" // LIBC_INLINE
 
-#include <fenv.h>
+#include "hdr/fenv_macros.h"
```
Comment (lntue) [inline libc/src/__support/FPUtil/rounding_mode.h:14]: Group the include headers together and sort them.

File: libc/src/__support/FPUtil/riscv/FEnvImpl.h [hunk 9]
```diff
@@ -13,7 +13,7 @@
 #include "src/__support/macros/attributes.h" // For LIBC_INLINE_ASM
 #include "src/__support/macros/config.h"     // For LIBC_INLINE
 
-#include <fenv.h>
+#include "hdr/fenv_macros.h"
```
Comment (lntue) [inline libc/src/__support/FPUtil/riscv/FEnvImpl.h:16]: Move this to the header group above.

File: libc/src/fenv/fegetenv.h [hunk 10]
```diff
@@ -9,7 +9,7 @@
 #ifndef LLVM_LIBC_SRC_FENV_FEGETENV_H
 #define LLVM_LIBC_SRC_FENV_FEGETENV_H
 
-#include <fenv.h>
+#include "hdr/fenv_macros.h"
```
Comment (lntue) [inline libc/src/fenv/fegetenv.h:12]: These functions also use some type definitions from the system header `fenv.h`, not the macros, so you cannot replace the `#include <fenv.h>` here yet.  Same with the implementations.

File: libc/src/__support/FPUtil/FEnvImpl.h [hunk 11]
```diff
@@ -9,11 +9,11 @@
 #ifndef LLVM_LIBC_SRC___SUPPORT_FPUTIL_FENVIMPL_H
 #define LLVM_LIBC_SRC___SUPPORT_FPUTIL_FENVIMPL_H
 
+#include "hdr/fenv_macros.h"
 #include "hdr/math_macros.h"
 #include "src/__support/macros/attributes.h" // LIBC_INLINE
 #include "src/__support/macros/properties/architectures.h"
 #include "src/errno/libc_errno.h"
-#include <fenv.h>
```
Comment (lntue) [inline libc/src/__support/FPUtil/FEnvImpl.h:16]: We still need to keep this `#include <fenv.h>` for the `fenv_t` definition in `get_env` and `set_env` below.  This would be remove in the next step when we add proxy header for `fenv_t`.

File: libc/src/__support/FPUtil/aarch64/FEnvImpl.h [hunk 12]
```diff
@@ -17,9 +17,9 @@
 #endif
 
 #include <arm_acle.h>
-#include <fenv.h>
```
Comment (lntue) [inline libc/src/__support/FPUtil/aarch64/FEnvImpl.h:20]: ditto

File: libc/src/__support/FPUtil/aarch64/fenv_darwin_impl.h [hunk 13]
```diff
@@ -17,9 +17,9 @@
 #endif
 
 #include <arm_acle.h>
-#include <fenv.h>
```
Comment (lntue) [inline libc/src/__support/FPUtil/aarch64/fenv_darwin_impl.h:20]: ditto

File: libc/src/__support/FPUtil/arm/FEnvImpl.h [hunk 14]
```diff
@@ -9,10 +9,9 @@
 #ifndef LLVM_LIBC_SRC___SUPPORT_FPUTIL_ARM_FENVIMPL_H
 #define LLVM_LIBC_SRC___SUPPORT_FPUTIL_ARM_FENVIMPL_H
 
+#include "hdr/fenv_macros.h"
 #include "src/__support/FPUtil/FPBits.h"
 #include "src/__support/macros/attributes.h" // For LIBC_INLINE
-
-#include <fenv.h>
```
Comment (lntue) [inline libc/src/__support/FPUtil/arm/FEnvImpl.h:15]: ditto

File: libc/src/__support/FPUtil/CMakeLists.txt [hunk 15]
```diff
@@ -3,7 +3,7 @@ add_header_library(
   HDRS
     FEnvImpl.h
   DEPENDS
-    libc.include.fenv
```
Comment (lntue) [inline libc/src/__support/FPUtil/CMakeLists.txt:6]: We still need to have `libc.include.fenv` in the dependency for this target for now.  We can remove it once `FEnvImpl.h` does not need `#include <fenv.h>` anymore.

File: libc/src/__support/FPUtil/riscv/FEnvImpl.h [hunk 16]
```diff
@@ -9,11 +9,10 @@
 #ifndef LLVM_LIBC_SRC___SUPPORT_FPUTIL_RISCV_FENVIMPL_H
 #define LLVM_LIBC_SRC___SUPPORT_FPUTIL_RISCV_FENVIMPL_H
 
+#include "hdr/fenv_macros.h"
 #include "src/__support/FPUtil/FPBits.h"
 #include "src/__support/macros/attributes.h" // For LIBC_INLINE_ASM
 #include "src/__support/macros/config.h"     // For LIBC_INLINE
-
-#include <fenv.h>
```
Comment (lntue) [inline libc/src/__support/FPUtil/riscv/FEnvImpl.h:16]: ditto

File: libc/src/__support/FPUtil/x86_64/FEnvImpl.h [hunk 17]
```diff
@@ -16,7 +16,7 @@
 #error "Invalid include"
 #endif
 
-#include <fenv.h>
```
Comment (lntue) [inline libc/src/__support/FPUtil/x86_64/FEnvImpl.h:19]: ditto

File: libc/test/UnitTest/FPExceptMatcher.cpp [hunk 18]
```diff
@@ -8,7 +8,7 @@
 
 #include "FPExceptMatcher.h"
 
-#include <fenv.h>
```
Comment (lntue) [inline libc/test/UnitTest/FPExceptMatcher.cpp:11]: Don't replace the include for this file yet, it only uses `fenv_t` and does not use any of the macro constants from `fenv.h`.

File: libc/test/src/fenv/exception_flags_test.cpp [hunk 19]
```diff
@@ -12,7 +12,7 @@
 #include "src/__support/FPUtil/FEnvImpl.h"
 #include "test/UnitTest/Test.h"
 
-#include <fenv.h>
```
Comment (lntue) [inline libc/test/src/fenv/exception_flags_test.cpp:15]: This test uses `fexcept_t`, so it still needs `#include <fenv.h>`.  We will replace it in the next step when we add proxy header for `fexcept_t`.

File: libc/test/src/fenv/feholdexcept_test.cpp [hunk 20]
```diff
@@ -13,7 +13,7 @@
 #include "test/UnitTest/FPExceptMatcher.h"
 #include "test/UnitTest/Test.h"
 
-#include <fenv.h>
```
Comment (lntue) [inline libc/test/src/fenv/feholdexcept_test.cpp:16]: This test uses `fenv_t`, so it still needs #include <fenv.h>. We will replace it in the next step when we add proxy header for `fenv_t`.

File: libc/test/src/fenv/feupdateenv_test.cpp [hunk 21]
```diff
@@ -11,7 +11,7 @@
 #include "src/__support/FPUtil/FEnvImpl.h"
 #include "test/UnitTest/Test.h"
 
-#include <fenv.h>
```
Comment (lntue) [inline libc/test/src/fenv/feupdateenv_test.cpp:14]: This test uses `fenv_t`, so it still needs #include <fenv.h>. We will replace it in the next step when we add proxy header for `fenv_t`.

File: libc/test/src/fenv/getenv_and_setenv_test.cpp [hunk 22]
```diff
@@ -14,7 +14,7 @@
 #include "src/__support/FPUtil/FEnvImpl.h"
 #include "test/UnitTest/Test.h"
 
-#include <fenv.h>
```
Comment (lntue) [inline libc/test/src/fenv/getenv_and_setenv_test.cpp:17]: This test uses `fenv_t`, so it still needs #include <fenv.h>. We will replace it in the next step when we add proxy header for `fenv_t`.

File: libc/utils/MPFRWrapper/MPFRUtils.cpp [hunk 23]
```diff
@@ -14,8 +14,8 @@
 #include "src/__support/FPUtil/fpbits_str.h"
 #include "test/UnitTest/FPMatcher.h"
 
+#include "hdr/fenv_macros.h"
```
Comment (lntue) [inline libc/utils/MPFRWrapper/MPFRUtils.cpp:17]: Look like this is not used directly in this file.  You can remove it.

File: libc/AOR_v20.02/math/test/mathtest.c [hunk 24]
```diff
@@ -6,17 +6,17 @@
  * SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
  */
 
+#include "mathlib.h"
```
Comment (lntue) [inline libc/AOR_v20.02/math/test/mathtest.c:9]: Revert the changes in this file.

File: libc/AOR_v20.02/math/test/ulp.c [hunk 25]
```diff
@@ -6,6 +6,7 @@
  * SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
  */
 
+#include "mathlib.h"
```
Comment (lntue) [inline libc/AOR_v20.02/math/test/ulp.c:9]: Revert the changes in this file.

File: libc/hdr/CMakeLists.txt [hunk 26]
```diff
@@ -31,3 +31,12 @@ add_proxy_header_library(
...
     libc.include.math
 )
+
+add_proxy_header_library(
+  fenv_macros
+  HDRS
+    fenv_macros.h
+  FULL_BUILD_DEPENDS
+    libc.include.llvm-libc-macros.fenv_macros
+    libc.hdr.fenv_macros
```
Comment (lntue) [inline libc/hdr/CMakeLists.txt:41]: This should be `libc.include.fenv`

File: libc/src/__support/FPUtil/aarch64/FEnvImpl.h [hunk 27]
```diff
@@ -17,10 +17,10 @@
...
 #include <stdint.h>
 
 #include "src/__support/FPUtil/FPBits.h"
+#include <fenv.h>
```
Comment (lntue) [inline libc/src/__support/FPUtil/aarch64/FEnvImpl.h:23]: Keep `#include <fenv.h>` in where it was, and add `#include "hdr/fenv_macros.h` here also.  It seems redundant, but it will make the next step easier.

File: libc/src/__support/FPUtil/CMakeLists.txt [hunk 28]
```diff
@@ -3,7 +3,7 @@ add_header_library(
...
+    libc.include.fenv
```
Comment (lntue) [inline libc/src/__support/FPUtil/CMakeLists.txt:6]: also add `libc.hdr.fenv_macros` dependency here.

File: libc/src/__support/FPUtil/aarch64/fenv_darwin_impl.h [hunk 29]
```diff
@@ -17,10 +17,10 @@
...
 #include <stdint.h>
 
 #include "src/__support/FPUtil/FPBits.h"
+#include <fenv.h>
```
Comment (lntue) [inline libc/src/__support/FPUtil/aarch64/fenv_darwin_impl.h:23]: Keep #include <fenv.h> in where it was, and add #include "hdr/fenv_macros.h here also. It seems redundant, but it will make the next step easier.

File: libc/src/__support/FPUtil/arm/FEnvImpl.h [hunk 30]
```diff
@@ -11,7 +11,6 @@
 
 #include "src/__support/FPUtil/FPBits.h"
```
Comment (lntue) [inline libc/src/__support/FPUtil/arm/FEnvImpl.h:12]: Add #include "hdr/fenv_macros.h here also. It seems redundant, but it will make the next step easier.

File: libc/src/__support/FPUtil/riscv/FEnvImpl.h [hunk 31]
```diff
@@ -12,7 +12,6 @@
 #include "src/__support/FPUtil/FPBits.h"
```
Comment (lntue) [inline libc/src/__support/FPUtil/riscv/FEnvImpl.h:12]: Add `#include "hdr/fenv_macros.h"` here also. It seems redundant, but it will make the next step easier.

File: libc/src/fenv/CMakeLists.txt [hunk 32]
```diff
@@ -5,7 +5,7 @@ add_entrypoint_object(
   HDRS
     fegetround.h
   DEPENDS
-    libc.include.fenv
+    libc.hdr.fenv_macros
```
Comment (lntue) [inline libc/src/fenv/CMakeLists.txt:8]: Revert dependency changes in this file.

File: utils/bazel/llvm-project-overlay/libc/BUILD.bazel [hunk 33]
```diff
@@ -70,6 +70,12 @@ libc_support_library(
     deps = [":llvm_libc_macros_limits_macros"],
 )
 
+libc_support_library(
+    name = "llvm_libc_macros_fenv_macros",
+    hdrs = ["include/llvm-libc-macros/fenv-macros.h"],
+    deps = [":llvm_libc_macros_limits_macros"],
+)
+
```
Comment (lntue) [inline utils/bazel/llvm-project-overlay/libc/BUILD.bazel:78]: You don't need to add this one.

File: utils/bazel/MODULE.bazel
Comment (lntue) [inline utils/bazel/MODULE.bazel:1]: Remove this file from the PR

File: utils/bazel/MODULE.bazel.lock
Comment (lntue) [inline utils/bazel/MODULE.bazel.lock:1]: Remove this file from the PR.

File: libc/AOR_v20.02/math/test/mathtest.c
Comment (lntue) [inline libc/AOR_v20.02/math/test/mathtest.c:1]: You can just revert this file to its current version in main.

File: libc/AOR_v20.02/math/test/ulp.c
Comment (lntue) [inline libc/AOR_v20.02/math/test/ulp.c:1]: You can just revert this file to its current version in main.

**Comment by lntue** [comment]
LGTM overall for me.  Let see if others have any further comments.

File: libc/src/__support/FPUtil/aarch64/FEnvImpl.h (hunk 27 above)
Comment (nickdesaulniers) [reply libc/src/__support/FPUtil/aarch64/FEnvImpl.h:24]: This line should not be moved, please move it back.

File: libc/src/__support/FPUtil/aarch64/fenv_darwin_impl.h (hunk 29 above)
Comment (nickdesaulniers) [reply libc/src/__support/FPUtil/aarch64/fenv_darwin_impl.h:24]: same

File: libc/src/__support/FPUtil/aarch64/FEnvImpl.h (hunk 27 above)
Comment (nickdesaulniers) [reply libc/src/__support/FPUtil/aarch64/FEnvImpl.h:24]: actually, most files remove the include of `fenv.h` entirely. Any idea what we're keeping it around for here?

File: libc/src/__support/FPUtil/aarch64/fenv_darwin_impl.h (hunk 29 above)
Comment (nickdesaulniers) [reply libc/src/__support/FPUtil/aarch64/fenv_darwin_impl.h:24]: actually, most files remove the include of `fenv.h` entirely. Any idea what we're keeping it around for here?

Comment (Jobhdez) [reply libc/src/__support/FPUtil/aarch64/fenv_darwin_impl.h:24]: intue said that this will make the next step easier to do. I think in the next PR I am going to make a proxy header for the fenv types so maybe thats why.

Comment (nickdesaulniers) [reply libc/src/__support/FPUtil/aarch64/fenv_darwin_impl.h:24]: Ok, the point about not moving the include still stands though. Just undo that change please.

**Review by nickdesaulniers — APPROVED** [review]
thanks for the patch!

**Review by lntue — APPROVED** [review]

**Comment by Jobhdez** [comment]
> thanks for the patch!

no problem!

File: libc/hdr/CMakeLists.txt [hunk 34]
```diff
@@ -31,3 +31,12 @@ add_proxy_header_library(
...
     libc.include.math
 )
+
+add_proxy_header_library(
+  fenv_macros
+  HDRS
+    fenv_macros.h
+  FULL_BUILD_DEPENDS
+    libc.include.llvm-libc-macros.fenv_macros
+    libc.incude.fenv
```
Comment (nickdesaulniers) [inline libc/hdr/CMakeLists.txt:41]: typo: s/incude/include/

Comment (nickdesaulniers) [reply libc/hdr/CMakeLists.txt:41]: fixed in 8cfa72ade9f2

Comment (lntue) [reply libc/hdr/CMakeLists.txt:41]: Thanks!
